src/pulselistener/pulselistener/mercurial.py

```diff
@@ -0,0 +1,148 @@
+# -*- coding: utf-8 -*-
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import asyncio
+import atexit
+import io
+import os
+import tempfile
+
+import hglib
+
+from cli_common.log import get_logger
+from cli_common.mercurial import robust_checkout
+from pulselistener.config import REPO_TRY
+
+logger = get_logger(__name__)
+
+
+class MercurialWorker(object):
+    '''
+    Mercurial worker maintaining a local clone of Mozilla-Central
+    '''
+    def __init__(self, phabricator_api, ssh_user, ssh_key, repo_url, repo_dir):
+        self.repo_url = repo_url
+        self.repo_dir = repo_dir
+        self.phabricator_api = phabricator_api
+
+        # Build asyncio shared queue
+        self.queue = asyncio.Queue()
+
+        # Write ssh key from secret
+        _, self.ssh_key_path = tempfile.mkstemp(suffix='.key')
+        with open(self.ssh_key_path, 'w') as f:
+            f.write(ssh_key)
+
+        # Build ssh conf
+        conf = {
+            'StrictHostKeyChecking': 'no',
+            'User': ssh_user,
+            'IdentityFile': self.ssh_key_path,
+        }
+        self.ssh_conf = 'ssh {}'.format(' '.join('-o {}="{}"'.format(k, v) for k, v in conf.items())).encode('utf-8')
+
+        # Remove key when finished
+        atexit.register(self.cleanup)
+
+    def cleanup(self):
+        os.unlink(self.ssh_key_path)
+        logger.info('Removed ssh key')
+
+    async def run(self):
+        # Start by updating the repo
+        logger.info('Checking out tip', repo=self.repo_url)
+        self.repo = robust_checkout(self.repo_url, self.repo_dir)
+        self.repo.setcbout(lambda msg: logger.info('Mercurial', stdout=msg))
+        self.repo.setcberr(lambda msg: logger.info('Mercurial', stderr=msg))
+        logger.info('Initial clone finished')
+
+        # Wait for phabricator diffs to apply
+        while True:
```

Comment:
We should pull the changes from mozilla-unified, otherwise our local clone will get out of date.

Reply:
Good idea, i was thinking of updating every X minutes in a later commit, but that may be overkill for now

Reply:
We could also update once every X revisions, which is an approximation of once every X minutes and might be easier to implement.

Reply:
I simply added the update on each diff, it's pretty fast as it keeps getting updated. We'll see how it behaves once on Heroku